### PR TITLE
Fix an issue of manual / auto compaction data race

### PR DESCRIPTION
**Pull request by siying**
Summary: A data race between a manual and an auto compaction can cause a scheduled automatic compaction to be cancelled and never rescheduled again. This may cause a condition of hanging forever. Fix this by always making sure the cancelled compaction is put back to the compaction queue.

Test Plan:Add a unit test.

### CODE REVIEW

File: db/db_compaction_test.cc [hunk 1]
```diff
@@ -1852,6 +1852,47 @@ TEST_F(DBCompactionTest, L0_CompactionBug_Issue44_b) {
...
+  Flush();
+  Put("foo", "");
+  Put("bar", "");
+  Flush();
+
+  // The auto compaction is scheduled but waited until here
+  TEST_SYNC_POINT("DBCompactionTest::ManualAutoRace:1");
+  // The auto compaction will wait until the the manual compaction is registerd
+  // before processing so that it will be cancelled.
+  dbfull()->CompactRange(CompactRangeOptions(), handles_[1], nullptr, nullptr);
```
Comment (ajkr) [inline db/db_compaction_test.cc]: if you dynamically set disable_auto_compactions=true here (and then unset it later), can you then verify that the compaction in CF 0 was canceled? or is that outside the scope of this test?

Comment (siying) [inline db/db_compaction_test.cc]: I don't know but it feels like it invokes a different code path from what I want to cover.

Comment (ajkr) [inline db/db_compaction_test.cc]: sorry I meant set disable_auto_compactions=true after the first auto-compaction starts but before the second can be scheduled. Then you can verify the first one is really canceled.

this is not a blocking comment, I think your test case is right, so feel free to commit as-is.

Comment (siying) [inline db/db_compaction_test.cc]: Let me commit it for now. We can continue the discussion that how to verify that compaction has been cancelled.

**Pull request closed without merging**